src/Compilers/CSharp/Portable/BoundTree/TupleBinaryOperatorInfo.cs

```diff
@@ -0,0 +1,108 @@
+        internal string Dump() => TreeDumper.DumpCompact(DumpCore());
+#endif
+
+        private TupleBinaryOperatorInfo(TypeSymbol leftConvertedType, TypeSymbol rightConvertedType)
+        {
+            LeftConvertedType = leftConvertedType;
+            RightConvertedType = rightConvertedType;
+        }
+
+        /// <summary>
+        /// Holds the information for an element-wise comparison (like `a == b`)
+        /// </summary>
+        internal class Single : TupleBinaryOperatorInfo
+        {
+            internal readonly BinaryOperatorKind Kind;
+            internal readonly MethodSymbol MethodSymbolOpt; // User-defined comparison operator, if applicable
+
+            // To convert the result of comparison to bool
+            internal readonly Conversion BoolConversion;
+            internal readonly UnaryOperatorSignature BoolOperator; // Information for op_true or op_false
+
+            internal Single(TypeSymbol leftConvertedType, TypeSymbol rightConvertedType, BinaryOperatorKind kind,
+                MethodSymbol methodSymbolOpt, Conversion boolConversion, UnaryOperatorSignature boolOperator) : base(leftConvertedType, rightConvertedType)
+            {
+                Kind = kind;
+                MethodSymbolOpt = methodSymbolOpt;
+                BoolConversion = boolConversion;
+                BoolOperator = boolOperator;
+
+                // If a user-defined comparison operator is present, then the operator kind must be user-defined
+                Debug.Assert(Kind.IsUserDefined() || (object)MethodSymbolOpt == null);
+            }
+
+            internal override bool IsSingle()
+                => true;
+
+            public override string ToString()
+                => $"binaryOperatorKind: {Kind}";
+
+#if DEBUG
+            internal override TreeDumperNode DumpCore()
+            {
+                var sub = new List<TreeDumperNode>();
+                if ((object)MethodSymbolOpt != null)
+                {
+                    sub.Add(new TreeDumperNode("methodSymbolOpt", MethodSymbolOpt.ToDisplayString(), null));
+                }
+                sub.Add(new TreeDumperNode("leftConversion", LeftConvertedType.ToDisplayString(), null));
+                sub.Add(new TreeDumperNode("rightConversion", RightConvertedType.ToDisplayString(), null));
+
+                return new TreeDumperNode("nested", Kind, sub);
+            }
+#endif
+        }
+
+        /// <summary>
+        /// Holds the information for a tuple comparison, either at the top-level (like `(a, b) == ...`) or nested (like `(..., (a, b)) == (..., ...)`).
+        /// </summary>
+        internal class Multiple : TupleBinaryOperatorInfo
+        {
+            internal readonly ImmutableArray<TupleBinaryOperatorInfo> Operators;
+
+            internal Multiple(ImmutableArray<TupleBinaryOperatorInfo> operators, TypeSymbol leftConvertedType, TypeSymbol rightConvertedType) 
+                : base(leftConvertedType, rightConvertedType)
```

Comment:
>leftConvertedType, rightConvertedTyp [](start = 23, length = 36)

Can these types be different if we ignore tuple names? Perhaps Assert if should match?

Reply:
Yes, these types can be different. If you implement an operator==(A, B), then `(a, a) == (b, b)` will have converted type `(A, A)` on the left and `(B, B)` on the right.

---
In reply to: [170494879](https://github.com/dotnet/roslyn/pull/23356#discussion_r170494879) [](ancestors = 170494879)